File: frontend/src/types/agent.ts
/**
 * Agent API Types
 */

export interface ToolCall {
  tool: string;
  input: Record<string, any>;
  result: any;
}

export interface AgentResponse {
  response: string;
  agent_name: string;
  model?: string;
  tool_calls: ToolCall[];
  iterations: number;
  routing_info?: Record<string, any>;
  response_data?: Record<string, any>;
}

export interface ChatMessage {
  id?: string;
  role: 'user' | 'assistant';
  content: string;
  timestamp: string;
  agent_name?: string;
  tool_calls?: ToolCall[];
  response_data?: Record<string, any>;
}

/**
 * 대화 이력 메시지 (API 전송용)
 */
export interface ConversationMessage {
  role: 'user' | 'assistant';
  content: string;
}

export interface AgentRequest {
  message: string;
  context?: Record<string, any>;
  tenant_id?: string;
  conversation_history?: ConversationMessage[];
}

// SSE Streaming Event Types
export type SSEEventType = 'start' | 'routing' | 'routed' | 'processing' | 'content' | 'tools' | 'workflow' | 'response_data' | 'done' | 'error';

// 워크플로우 노드 타입 (재귀적 정의)
export interface WorkflowNode {
  id: string;
  type: 'condition' | 'action' | 'if_else' | 'loop' | 'parallel';
  config: Record<string, unknown>;
  next?: string[];
  then_nodes?: WorkflowNode[];
  else_nodes?: WorkflowNode[];
  loop_nodes?: WorkflowNode[];
  parallel_nodes?: WorkflowNode[];
}

// 워크플로우 DSL 타입 (SSE 이벤트용)
export interface WorkflowDSL {
  name?: string;
  description?: string;
  trigger: {
    type: 'event' | 'schedule' | 'manual';
    config?: Record<string, unknown>;
  };
  nodes: WorkflowNode[];
}

export interface SSEEvent {
  type: SSEEventType;
  message?: string;
  agent?: string;
  content?: string;
  tool_calls?: Array<{ tool: string; input: Record<string, any> }>;
  agent_name?: string;
  model?: string;
  iterations?: number;
  // 워크플로우 이벤트 관련 필드
  workflow?: {
    dsl: WorkflowDSL;
    workflowId?: string;
    workflowName?: string;
  };
  // response_data 이벤트 관련 필드 (BI 인사이트, 차트, 테이블 등)
  data?: Record<string, any>;
}

export interface StreamingState {
  isStreaming: boolean;
  currentAgent?: string;
  status?: string;
}

File: backend/app/schemas/agent.py
"""
Agent API Pydantic Schemas
"""
from typing import Any, Dict, List, Literal, Optional
from pydantic import BaseModel, Field


class ChatMessage(BaseModel):
    """채팅 메시지"""

    role: str = Field(..., description="메시지 역할 (user, assistant)")
    content: str = Field(..., description="메시지 내용")


class ConversationMessage(BaseModel):
    """대화 이력의 개별 메시지 (타입 검증 강화)"""

    role: Literal["user", "assistant"] = Field(..., description="메시지 역할")
    content: str = Field(..., description="메시지 내용")


class AgentRequest(BaseModel):
    """Agent 실행 요청"""

    message: str = Field(..., description="사용자 메시지")
    context: Optional[Dict[str, Any]] = Field(default=None, description="추가 컨텍스트")
    tenant_id: Optional[str] = Field(default=None, description="테넌트 ID")
    conversation_history: Optional[List[ConversationMessage]] = Field(
        default=None,
        description="대화 이력 (최근 대화 컨텍스트 유지용)"
    )


class ToolCall(BaseModel):
    """Tool 호출 정보"""

    tool: str = Field(..., description="Tool 이름")
    input: Dict[str, Any] = Field(..., description="Tool 입력")
    result: Any = Field(..., description="Tool 실행 결과")


class AgentResponse(BaseModel):
    """Agent 실행 응답"""

    response: str = Field(..., description="Agent 응답 메시지")
    agent_name: str = Field(..., description="실행된 Agent 이름")
    tool_calls: List[ToolCall] = Field(default_factory=list, description="Tool 호출 이력")
    iterations: int = Field(..., description="실행 반복 횟수")
    routing_info: Optional[Dict[str, Any]] = Field(
        default=None, description="Meta Router의 라우팅 정보"
    )

    class Config:
        from_attributes = True


class JudgmentRequest(BaseModel):
    """Judgment Agent 직접 호출 요청"""

    message: str = Field(..., description="판단 요청 메시지")
    sensor_data: Optional[Dict[str, Any]] = Field(
        default=None, description="센서 데이터"
    )
    ruleset_id: Optional[str] = Field(default=None, description="Ruleset ID")
    tenant_id: Optional[str] = Field(default=None, description="테넌트 ID")
